### PR TITLE
Add start to backtrace for debugger

### PR DESCRIPTION
**Pull request by Yuyupo**
With this you can add a start where you want to start listing the backtraces, not just maximum depth.

JerryScript-DCO-1.0-Signed-off-by: Daniella Barsony bella@inf.u-szeged.hu

### CODE REVIEW

File: jerry-core/debugger/debugger.h [hunk 1]
```diff
@@ -376,6 +376,7 @@ typedef struct
 typedef struct
 {
   uint8_t type; /**< type of the message */
+  uint8_t start; /**< starting point (0 - max_depth)*/
```
Comment (zherczeg) [inline jerry-core/debugger/debugger.h]: This should be:
`uint8_t min_depth[sizeof (uint32_t)]; /**< minimum depth */`

Similar to:
`uint8_t max_depth[sizeof (uint32_t)]; /**< maximum depth (0 - unlimited) */`

**Comment by Yuyupo** [comment]
@zherczeg Changed what was requested

File: jerry-debugger/jerry-client-ws.py [hunk 2]
```diff
@@ -348,22 +348,30 @@ def do_delete(self, args):
...
         max_depth = 0
+        min_depth = 0
 
         if args:
+            args = args.split(" ")
             try:
-                max_depth = int(args)
-                if max_depth <= 0:
+                min_depth = int(args[0])
+                max_depth = int(args[1])
```
Comment (LaszloLango) [inline jerry-debugger/jerry-client-ws.py]: What happens if there is` args` length is 1 and there is no second element?
Both `bt max_value` and `bt min_value max_value` commands should work.

Comment (LaszloLango) [reply jerry-debugger/jerry-client-ws.py]: I see now, this is intentional and the `bt` command has two mandatory arguments. IMHO this should be changed. One argument should mean a `max_value` and two arguments should mean `min_value max_value` pair.

Comment (zherczeg) [reply jerry-debugger/jerry-client-ws.py]: I don't mind having two mandatory arguments.

Comment (zherczeg) [reply jerry-debugger/jerry-client-ws.py]: Sorry I misunderstood you. Yes, in the python debugger, only one argument must be mandatory. In the message, both values must be sent.

File: tests/debugger/do_backtrace.expected [hunk 3]
```diff
@@ -14,10 +14,15 @@ out: function test
...
 Stopped at tests/debugger/do_backtrace.js:23 (in foo() at line:21, col:1)
+(jerry-debugger) bt 1 2
+Frame 0: tests/debugger/do_backtrace.js:23 (in foo() at line:21, col:1)
+Frame 1: tests/debugger/do_backtrace.js:33 (in test() at line:30, col:1)
 (jerry-debugger) bt
 Frame 0: tests/debugger/do_backtrace.js:23 (in foo() at line:21, col:1)
 Frame 1: tests/debugger/do_backtrace.js:33 (in test() at line:30, col:1)
 Frame 2: tests/debugger/do_backtrace.js:40
+(jerry-debugger) bt 1 1
+Frame 0: tests/debugger/do_backtrace.js:23 (in foo() at line:21, col:1)
```
Comment (zherczeg) [inline tests/debugger/do_backtrace.expected]: When you specify a min, then maximum (max-min) items could be sent. Here (1-1)==0, so the backtrace should be empty (empty backtrace is valid, you need to test it!!!)

So "bt 1 2" above should only send 1 item. Please also test a large number, e.g. "bt 1000 1001".

File: tests/debugger/do_backtrace.expected [hunk 4]
```diff
@@ -14,10 +14,15 @@ out: function test
 Stopped at tests/debugger/do_backtrace.js:33 (in test() at line:30, col:1)
 (jerry-debugger) s
 Stopped at tests/debugger/do_backtrace.js:23 (in foo() at line:21, col:1)
+(jerry-debugger) bt 1 2
+Frame 0: tests/debugger/do_backtrace.js:23 (in foo() at line:21, col:1)
```
Comment (LaszloLango) [inline tests/debugger/do_backtrace.expected]: Is this intentional? I thought this should be `Frame 1: tests/debugger/do_backtrace.js:33 (in test() at line:30, col:1)`

File: tests/debugger/do_backtrace.cmd [hunk 5]
```diff
@@ -4,9 +4,14 @@ next
...
+bt 1 2
 bt
+bt 2
 n
 n
 s
 backtrace
+bt 4 4
+bt 600 919
+bt 3 500
```
Comment (LaszloLango) [inline tests/debugger/do_backtrace.cmd]: What happens if `min` greater than the `max`? Can we test it?

**Comment by Yuyupo** [comment]
@LaszloLango Fixed what was requested, and added the test case that was missing.

File: tests/debugger/do_backtrace.expected [hunk 6]
```diff
@@ -14,10 +14,15 @@ out: function test
 Stopped at tests/debugger/do_backtrace.js:33 (in test() at line:30, col:1)
 (jerry-debugger) s
 Stopped at tests/debugger/do_backtrace.js:23 (in foo() at line:21, col:1)
+(jerry-debugger) bt 1 2
+Frame 0: tests/debugger/do_backtrace.js:33 (in test() at line:30, col:1)
```
Comment (LaszloLango) [inline tests/debugger/do_backtrace.expected]: `Frame 1`

File: tests/debugger/do_backtrace.expected [hunk 7]
```diff
@@ -14,10 +14,15 @@ out: function test
 Stopped at tests/debugger/do_backtrace.js:33 (in test() at line:30, col:1)
 (jerry-debugger) s
 Stopped at tests/debugger/do_backtrace.js:23 (in foo() at line:21, col:1)
+(jerry-debugger) bt 1 2
+Frame 1: tests/debugger/do_backtrace.js:23 (in foo() at line:21, col:1)
```
Comment (LaszloLango) [inline tests/debugger/do_backtrace.expected]: the frame index is still wrong

File: jerry-core/debugger/debugger.h [hunk 8]
```diff
@@ -376,6 +376,7 @@ typedef struct
 typedef struct
 {
   uint8_t type; /**< type of the message */
+  uint8_t min_depth[sizeof (uint32_t)]; /**< minimum depth*/
```
Comment (LaszloLango) [inline jerry-core/debugger/debugger.h]: Why don't we use `uint16_t` for both `min_depth` and `max_depth`? @zherczeg, what do you think?

Comment (zherczeg) [reply jerry-core/debugger/debugger.h]: Currently only uint8, cpointer, and uint32 types are used by the protocol. Adding new types would require adding new encoders/decoders for clients. This is certainly possible, but for this it does not seem worth the effort.

Comment (LaszloLango) [reply jerry-core/debugger/debugger.h]: I see. Thanks for the explanation.

File: jerry-debugger/jerry-client-ws.py [hunk 9]
```diff
@@ -348,22 +349,36 @@ def do_delete(self, args):
...
-                    print("Error: Positive integer number expected")
-                    return
+                if len(args) == 2:
+                    self.min_depth = int(args[0])
+                    max_depth = int(args[1])
+                    if max_depth <= 0 or self.min_depth < 0:
+                        print("Error: Positive integer number expected")
+                        return
+                    if self.min_depth > max_depth:
+                        print("Error: Start needs to be lower than depth")
```
Comment (zherczeg) [inline jerry-debugger/jerry-client-ws.py]: `Error: start depth needs to be less equal than max depth`

**Comment by Yuyupo** [comment]
@zherczeg Fixed that minor issue

**Pull request merged**